Libplanet/Net/Protocols/KademliaProtocol.cs

```diff
@@ -617,6 +678,111 @@ private void ReceivePing(Ping ping)
+
+            Task[] awaitables = peers.Select(peer =>
+                PingAsync(peer, _requestTimeout, cancellationToken)
+            ).ToArray();
+            try
+            {
+                await Task.WhenAll(awaitables);
+            }
+            catch (AggregateException e)
+            {
+                if (e.InnerExceptions.All(ie => ie is TimeoutException) &&
+                    e.InnerExceptions.Count == awaitables.Length)
+                {
+                    throw new TimeoutException(
+                        $"All neighbors found do not respond in {_requestTimeout}."
+                    );
+                }
+
+                _logger.Error(
+                    e,
+                    "Some responses from neighbors found unexpectedly terminated: {Exception}",
+                    e
+                );
+            }
+
+            for (int i = 0; i < closestNeighbors.Count; i++)
+            {
+                if (string.CompareOrdinal(
+                        closestNeighbors[i].Address.ToHex(),
+                        searchAddress.ToHex()
+                   ) == 0)
+                {
+                    if (_routing.Contains(closestNeighbors[i]))
+                    {
+                        peerFound = closestNeighbors[i];
+                        return peerFound;
+                    }
+                }
+            }
+
+            var findNeighboursTasks = new List<Task<BoundPeer>>();
+            Peer closestKnown = closestNeighbors.Count == 0 ? null : closestNeighbors[0];
+            for (int i = 0; i < Kademlia.FindConcurrency && i < peers.Count; i++)
+            {
+                if (peerFound is null ||
+                    string.CompareOrdinal(
+                        Kademlia.CalculateDistance(peers[i].Address, target).ToHex(),
+                        Kademlia.CalculateDistance(closestKnown.Address, target).ToHex()
+                    ) < 1)
+                {
+                    findNeighboursTasks.Add(FindSpecificPeerAsync(
+                        history,
+                        target,
+                        peers[i],
+                        (depth == -1) ? depth : depth + 1,
+                        searchAddress,
+                        timeout,
+                        cancellationToken));
+                }
+            }
+
+            var foundSpecificPeer = new List<BoundPeer>();
+            try
+            {
+                foundSpecificPeer.AddRange(await Task.WhenAll(findNeighboursTasks));
+            }
+            catch (TimeoutException)
+            {
+                if (findNeighboursTasks.All(findPeerTask => findPeerTask.IsFaulted))
+                {
+                    throw new TimeoutException(
+                        "Timeout exception occurred during ProcessFoundForSpecificAsync().");
+                }
+            }
+
+            if (foundSpecificPeer.Count > 0)
+            {
+                peerFound = foundSpecificPeer[0];
```

Comment:
`foundSpecificPeer[0]` may not have same address with `target` because it may not exists on the network. How about compare with `target` before this statement and if it's different, just return `null` instead?